Make recorder.purge_entities require at least one entity filter value

karwosts opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

Previously, `recorder.purge_entities` called with no arguments would purge all entities. This PR makes this call raise an error instead. 

It was too easy for a user to accidentally purge his entire database using the old logic. Users who want to purge all entities can still use the `recorder.purge` service to achieve the same effect, and that service is more obvious how it behaves. 


## Proposed change

In #106125, it is established that if a user tries to call the `recorder.purge_entities` service with a device (or area) selected in the target picker, the service call will clear **all data** from the database, not only the entities from that device (or area), as might naively be expected.

This happens because the service does not handle devices or areas and ignores these keys, but just by selecting one, it defeats the mechanism that was attempting to prevent the user from accidentally clearing all data, which required them to select an entity from the target picker: 

> Note: The entity_id is only optional when used in automations.yaml or scripts.yaml . When using the UI to call this service then it is mandatory to specify at least one entity_id using the Target Picker or via YAML mode.

By rejecting the service call if an area or device is provided, this accidental total data loss can be avoided. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #106125
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/32376

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

karwosts committed:
Disallow passing devices and areas to purge_entities

frenck commented on homeassistant/components/recorder/services.py:
```diff
@@ -35,15 +36,31 @@
+            vol.Optional(ATTR_KEEP_DAYS, default=0): cv.positive_int,
+        }
+    )
+    .extend(cv.ENTITY_SERVICE_FIELDS)
```
This ain't an entity service, so that is where it goes wrong.

karwosts replied:
I would have switched the target selector for an entity selector, but then I don't know how to meet this requirement:

> Note: The entity_id is only optional when used in automations.yaml or scripts.yaml . When using the UI to call this service then it is mandatory to specify at least one entity_id using the Target Picker or via YAML mode.

An entity selector as a field won't require a user to pick an entity, unless it is marked required, in which case I think it's no longer optional in automations.

frenck reviewed (dismissed):
As this ain't an entity service, it shouldn't be using/extending the entity service fields in its schema.

Additionally, the service description in `services.yaml` has a target, which is thus invalid as well.

The provided solution, isn't a solution, as the wrong selector is still show in the UI.

karwosts committed:
Revert "Disallow passing devices and areas to purge_entities"

This reverts commit 70c20076a410a58f1e79f5cc8147857eea64b2eb.

karwosts committed:
switch target for entity_id

karwosts committed:
change schema

karwosts commented:
Ok took another swing at this. I reverted previous changes and tried replacing the target picker with an entity picker, (and replaced the entity_service fields).

I played around with it a bit and it seems to work, I am not 100% sure if it can be considered a breaking change or not as scripts/automation that have `entity_id` saved under `target:` still seem to pass to pass that entity_id correctly after this change, so those still work, even though their UI renders a little screwy in the editor until the `target:` section is removed and replaced with `data: entity_id:`.

I am unclear if this service supported templating the entity_id previously, as the `ENTITY_SERVICE_FIELDS` that it was extending seem to indicate some kind of support for jinja templates, though I'm not sure if that ever really worked or is considered to be supported, the documentation does not mention templates here. So that's something to think about.

One also possible concern is related to this comment in the doc:
> When using the UI to call this service then it is mandatory to specify at least one entity_id using the Target Picker or via YAML mode.

It is true that if in the UI you don't put anything in the entity_id field in devtools, the service call will not execute and complain about lack of entity id (as it is marked as required in the yaml). However if you add, and then delete and entity_id, the field is left defined as an empty array, which _does_ pass the required field check, and would proceed to purge all entities. I'm not sure if that's something that should be addressed.

karwosts commented:
unstale

bdraco commented on homeassistant/components/recorder/services.py:
```diff
@@ -37,13 +38,14 @@
 
 SERVICE_PURGE_ENTITIES_SCHEMA = vol.Schema(
     {
+        vol.Optional(ATTR_ENTITY_ID, default=[]): cv.entity_ids,
```
# Case 1 - No filter
    # - All entities included

If there are no values it will remove everything so maybe we need a one of here?

bdraco replied:
example https://github.com/home-assistant/core/blob/6dcfe861fdc1907cdd88909d4a88ab2de4029586/homeassistant/helpers/config_validation.py#L460

karwosts replied:
Isn't that the intended, documented behavior?

> Leaving all three parameters empty will result in all entities being selected for purging.
> Note: The entity_id is only optional when used in automations.yaml or scripts.yaml. When using the UI to call this service then it is mandatory to specify at least one entity_id using the Target Picker or via YAML mode.

This is kind of an odd case where we partially rely on the frontend to not allow the user to submit this case through the devtools/services, even though it's technically supported 😕

karwosts replied:
I can see how a behavior change for this `purge_entities` servicce might make sense (I don't know why `purge` and `purge_entities` both can target all entities), but my first instinct was not for a breaking change here.

bdraco replied:
I think a breaking change is desirable here to not make mistakes as they can already use the `purge` service to do everything

karwosts replied:
If we make the breaking change to require at least one key, can we drop this behavior as well?

> When using the UI to call this service then it is mandatory to specify at least one entity_id using the Target Picker or via YAML mode.

It seems an odd restriction, that user is disallowed to use the service devtool to purge an entity_glob or domain.

bdraco replied:
> If we make the breaking change to require at least one key, can we drop this behavior as well?
> 
> > When using the UI to call this service then it is mandatory to specify at least one entity_id using the Target Picker or via YAML mode.
> 
> It seems an odd restriction, that user is disallowed to use the service devtool to purge an entity_glob or domain.

I _think_ the answer is yes there.  They should be able to pick any combination of them together to build the filter as long as they pick at least one so it doesn't select everything

karwosts replied:
Hmm I'm seeing that passing `entity_id: []` is both satisfying the `has_at_least_one_key` check, but also still deleting all entities. I think I may want some kind of check that it has at least one key that is not an empty array.

karwosts committed:
dont accept purge_entities with no arguments

bdraco commented on homeassistant/components/recorder/services.py:
```diff
@@ -35,15 +36,28 @@
+            vol.Optional(ATTR_KEEP_DAYS, default=0): cv.positive_int,
+        }
+    ),
+    vol.Any(
+        vol.Schema({vol.Required(ATTR_ENTITY_ID): vol.IsTrue()}, extra=vol.ALLOW_EXTRA),
+        vol.Schema({vol.Required(ATTR_DOMAINS): vol.IsTrue()}, extra=vol.ALLOW_EXTRA),
+        vol.Schema(
+            {vol.Required(ATTR_ENTITY_GLOBS): vol.IsTrue()}, extra=vol.ALLOW_EXTRA
         ),
```
I was hoping we had something off the shelf in config_validation for this, but I didn't find anything

karwosts replied:
Yeah I didn't find either. I'm not that experienced with voluptuous so I'm not sure if this is the most elegant way to do this, but was the first thing I stumbled into that seemed to achieve what I was looking for.

bdraco replied:
You could write your own validator like https://github.com/home-assistant/core/blob/1d6ae01baab69957e1bf7f26fc00564a4de9df9c/homeassistant/helpers/config_validation.py#L166 but I'm not sure its going to be any better as its likely just as many lines.

karwosts committed:
Merge branch 'dev' of https://github.com/home-assistant/core into purge-entities-no-devices

karwosts committed:
fix pytest

karwosts commented:
Tests fixed and docs updated.

bdraco commented:
In flight for the next 9 hours. Will have to take a look tomorrow

bdraco committed:
Merge branch 'dev' into purge-entities-no-devices

bdraco reviewed (approved):
Tested, working as expected.  Thanks @karwosts 

<img width="1105" alt="Screenshot 2024-04-22 at 5 41 36 PM" src="https://github.com/home-assistant/core/assets/663432/1bd504be-5ee1-4691-88d5-5f934eced443">
<img width="1134" alt="Screenshot 2024-04-22 at 5 42 11 PM" src="https://github.com/home-assistant/core/assets/663432/97d10b00-c190-4bd9-b812-07ce280dad10">
